src/metabase/lib/js.cljs

```diff
@@ -465,6 +469,82 @@
+        stage          (lib.util/query-stage a-query stage-number)
+        vis-columns    (lib.metadata.calculation/visible-columns a-query stage-number stage options)
+        ret-columns    (lib.metadata.calculation/returned-columns a-query stage-number stage)]
+    (to-array (lib.equality/mark-selected-columns
+                a-query stage-number vis-columns ret-columns {:keep-join? true}))))
+
+(defn ^:export legacy-field-ref
+  "Given a column metadata from eg. [[fieldable-columns]], return it as a legacy JSON field ref."
+  [column]
+  (-> column
+      lib.core/ref
+      convert/->legacy-MBQL
+      (update 2 update-vals #(if (qualified-keyword? %)
+                               (u/qualified-name %)
+                               %))
+      clj->js))
+
+(defn- legacy-ref->pMBQL [legacy-ref]
+  (-> legacy-ref
+      (js->clj :keywordize-keys true)
+      (update 0 keyword)
+      convert/->pMBQL))
+
+(defn- ->column-or-ref [column]
+  (if-let [^js legacy-column (when (object? column) column)]
+    (if (.-field_ref legacy-column)
+      ;; Prefer the attached field_ref if provided.
+      (legacy-ref->pMBQL (.-field_ref legacy-column))
+      ;; Fall back to converting like metadata.
+      (js.metadata/parse-column legacy-column))
+    ;; It's already a :metadata/column map
+    column))
+
+(defn ^:export find-column-indexes-from-legacy-refs
+  "Given a list of columns (either JS `data.cols` or MLv2 `ColumnMetadata`) and a list of legacy refs, find each ref's
+  corresponding index into the list of columns.
+
+  Returns a parallel list to the refs, with the corresponding index, or -1 if no matching column is found."
+  [a-query stage-number legacy-columns legacy-refs]
+  ;; Set up this query stage's `:aggregation` list as the context for [[convert/->pMBQL]] to convert legacy
+  ;; `[:aggregation 0]` refs into pMBQL `[:aggregation uuid]` refs.
+  (convert/with-aggregation-list (:aggregation (lib.util/query-stage a-query stage-number))
+    (let [columns       (mapv ->column-or-ref legacy-columns)
+          field-refs    (map legacy-ref->pMBQL legacy-refs)
+          ;; matches is a map of columns to the corresponding index in field-refs.
+          matches       (lib.equality/find-closest-matches-for-refs a-query stage-number field-refs columns {:keep-join? true})
+          ;; We want to return a parallel list to field-refs, giving the index of the matching column (or -1).
+          ;; First, map each column to its index (in the column list).
+          column->index (into {} (for [index (range (count columns))]
+                                   [(nth columns index) index]))
+          ;; And use that to map each match's ref-index to its column-index.
+          by-index      (into {} (for [[column ref-index] matches]
+                                   [ref-index (column->index column)]))
+          res (->> (range (count legacy-refs))
+                   (map #(by-index % -1))
+                   to-array)]
+      res)))
```

Comment:
Functions like this should go in the `cljc` files so they're easy to work with and test from a Clojure REPL, stuff in `js.cljs` should just be exports and light type conversion.

Reply:
I'm not sure I agree.

Pushing this logic out to one place isn't so bad. But search for everywhere that `metabase.lib.convert/legacy-ref->pMBQL` is called - that shows dealing with legacy refs escaping from `metabase.lib.js` and finding its way into `lib.field`, `lib.filter`, and one or two more. That feels to me like something that should be dealt with at the edges, not throughout MLv2.

On the other hand, this is mixing JS conversion with some massaging of the return type of `find-closest-matches-for-refs`. I could see moving that inner part into `lib.equality` and adding it to the `discouraged-vars` linter so no one else calls it. That would leave just the JS (and `legacy-ref->pMBQL`) conversions here.

Reply:
I've moved the guts of this to `lib.equality` and set that function as a `:discouraged-var`.